macros/src/main/scala/org/zalando/grafter/macros/DefaultReaderMacro.scala

```diff
@@ -1,31 +1,39 @@
 package org.zalando.grafter.macros
 
 import scala.annotation.StaticAnnotation
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox
 import ReaderMacros._
 
-import scala.meta._
-import DefaultReaderMacro._
-
 object DefaultReaderMacro {
 
   val annotationName = "defaultReader"
 
-  def expand(traitDef: Defn.Trait, objectDef: Option[Defn.Object], typeParam: Type.Name): Term.Block = {
-    output(traitDef, objectDef) {
-      q"""
-         implicit def reader[A](implicit defaultReader: cats.data.Reader[A, $typeParam]): cats.data.Reader[A, ${traitDef.name}] =
+  def impl(c: whitebox.Context)(annottees: c.Expr[Any]*): c.Expr[Any] = {
+    import c.universe._
+
+    val (classTree, companionTree): (Tree, Option[Tree]) =
+      annotationInputs(c)(annotationName)(annottees)
+
+    classTree match {
```

Comment:
We used to invoke `c.typecheck` on the `classTree` here to get the full types of the parameters, in order to deduplicate them later. Unfortunately the typechecking was throwing exceptions, then not having it would throw a NPE in `deduplicateParams`. I have now simplified that method and everything works fine.